tos/apps.py

```diff
@@ -0,0 +1,28 @@
+from django.apps import AppConfig
+from django.conf import settings
+from django.core.cache import caches
+from django.db.models.signals import pre_save
+from django.dispatch import receiver
+
+
+MIDDLEWARES = getattr(settings, 'MIDDLEWARE_CLASSES', [])
+
+
+class TOSConfig(AppConfig):
+    name = 'tos'
+    verbose_name = 'Terms Of Service'
+
+    def ready(self):
+        if 'tos.middleware.UserAgreementMiddleware' in MIDDLEWARES:
+            # Force the user to create a separate cache
+            cache = caches[getattr(settings, 'TOS_CACHE_NAME')]
+
+            TermsOfService = self.get_model('TermsOfService')
+
+            @receiver(pre_save, sender=TermsOfService, dispatch_uid='invalidate_cached_agreements')
+            def invalidate_cached_agreements(TermsOfService, instance, **kwargs):
+                if kwargs.get('raw', False):
+                    return
+
+                # Efficiently clear all keys from the cache
+                cache.clear()
```

Comment:
Great idea! I was clearing the cache previously because Django doesn't have a way of (agnostically) getting all cache keys that match a pattern, but this option is much better.

I've added a commit to the PR that does exactly this. Django even supports [cache versioning](https://docs.djangoproject.com/en/1.9/topics/cache/#cache-versioning) out of the box, so I just use that.

Note that this change does mean that there will be keys in the cache that won't ever be deleted. I've updated the documentation to discuss this.